src/pudl/metadata/classes.py

```diff
@@ -2028,6 +2045,41 @@ def sort_resource_names(resource: Resource):
 
         return sorted(resources, key=sort_resource_names, reverse=False)
 
+    @cached_property
+    def encoders(self) -> dict[SnakeCase, Encoder]:
+        """Compile a mapping of field names to their encoders, if they exist.
+
+        This dictionary will be used many times, so it makes sense to build it once
+        when the Package is instantiated so it can be reused.
+        """
+        encoded_fields = [
+            field
+            for res in self.resources
+            for field in res.schema.fields
+            if field.encoder
+        ]
+        encoders: dict[SnakeCase, Encoder] = {}
+        for field in encoded_fields:
+            if field.name not in encoders:
+                encoders[field.name] = field.encoder
+            else:
+                pd.testing.assert_frame_equal(encoders[field.name].df, field.encoder.df)
+                assert encoders[field.name].code_fixes == field.encoder.code_fixes
+                assert encoders[field.name].ignored_codes == field.encoder.ignored_codes
+        return encoders
+
+    def encode(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Clean up all coded columns in a dataframe based on PUDL coding tables.
+
+        Returns:
+            A modified copy of the input dataframe.
+        """
+        encoded_df = df.copy()
+        for col in df.columns:
+            if col in self.encoders:
+                encoded_df[col] = self.encoders[col].encode(df[col])
+        return encoded_df
```

Comment:
complete nit but it avoids a layer of nesting:
```suggestion
        encoded_df = df.copy()
        encoder_df_cols = [col for col in df if col in self.encoders]
        for encode_col in encoder_df_cols:
            encoded_df[col] = self.encoders[col].encode(df[col])
        return encoded_df
```

also are you meaning to copy to df as `encoded_df` but keep using `df` (in the column loop and in the `encode(df[col]`?

Reply:
I did intentionally leave `df` in the "source" role down below but that's being a little too clever. I'll swap it out.